GUI: Improve missing firmware handling

elad335 opened the pull request:
## Enhancements
* Install PS3UPDAT.PUP at the spot when booting games whenever firmware is missing. The option to boot games without firmware is still supported when all firmware SPRX are HLEd in firmware settings.
* Pop-up a confirmation dialog in firmware installation if firmware is already installed and log it.
## Screenshots:

![unknown (1)](https://user-images.githubusercontent.com/18193363/106036300-d30fd080-60dd-11eb-8718-e99230d1901e.png)
![image](https://user-images.githubusercontent.com/18193363/108876571-083d0f00-7607-11eb-916f-1385b28f8d0d.png)

Nekotekina commented on rpcs3/rpcs3qt/gui_settings.cpp:
```diff
@@ -165,34 +168,49 @@ void gui_settings::ShowBox(QMessageBox::Icon icon, const QString& title, const Q
+	if (clicked && clicked == on_install)
+	{
+		// Special path for firmware installation, parent must be main_window
+		dynamic_cast<main_window*>(parent)->InstallPup();
```
Why dynamic_cast?

elad335 replied:
Avoiding UB when misused, ensuring nullptr derference.

Nekotekina replied:
Which is a plain UB... Perhaps you could compare parent with known pointer instead.

elad335 replied:
Used emu callbacks instead.

Megamouse commented on rpcs3/Emu/RSX/RSXThread.cpp:
```diff
@@ -534,6 +534,8 @@ namespace rsx
 			return fmt::format("RSX [0x%07x]", +rsx->ctrl->get);
 		};
 
+		method_registers.init();
```
this commit has nothing to do with this PR

elad335 replied:
It unbreaks pure HLE mode.

Megamouse commented on rpcs3/rpcs3qt/gui_application.cpp:
```diff
@@ -354,6 +354,11 @@ void gui_application::InitializeCallbacks()
 
+	callbacks.on_install_fw = [this]()
+	{
+		return m_main_window->InstallPup();
```
this will crash in no-gui mode

elad335 replied:
But callback is not called in no-gui mode.

Megamouse replied:
doesn't matter where it's called from.

elad335 replied:
Fixed.

Megamouse commented on rpcs3/rpcs3qt/gui_settings.cpp:
```diff
@@ -5,7 +5,10 @@
 
 #include "Emu/System.h"
 
+#include "main_window.h"
```
why include this?

Megamouse commented on rpcs3/rpcs3qt/gui_settings.cpp:
```diff
@@ -5,7 +5,10 @@
+#include "main_window.h"
+
 #include <QCheckBox>
+#include <QPushButton>
```
why include this?

elad335 replied:
Needed by added code in pr.

Megamouse replied:
this is leftover code

Megamouse commented on rpcs3/rpcs3qt/gui_settings.cpp:
```diff
@@ -165,34 +168,49 @@ void gui_settings::ShowBox(QMessageBox::Icon icon, const QString& title, const Q
-
-	QMessageBox* mb = new QMessageBox(icon, title, text, buttons, parent, Qt::Dialog | Qt::MSWindowsFixedSizeDialogHint | (always_on_top ? Qt::WindowStaysOnTopHint : Qt::Widget));
-	mb->deleteLater();
+	QMessageBox* mb = new QMessageBox(icon, title, text, {}, parent, Qt::Dialog | Qt::MSWindowsFixedSizeDialogHint | (always_on_top ? Qt::WindowStaysOnTopHint : Qt::Widget));
```
This function is not supposed to know about different gui_saves, so why add installation specifics here?

elad335 replied:
Avoiding duplication of code with GUI settings interaction.

Megamouse replied:
There are multiple bad decisions here...
- You move from a simple QMessageBox::StandardButton implementation to a custom implementation that regresses the existing code and also adds more unnecessary translation
- You add a specific firmware install callback into this method

So my suggestion:
- Check the result param after calling ShowBox to call the Emu callback. This removes the nasty hack.
- Either use the ib_confirm_fw entry or not. And if you use it, don't add code based on entry.... you may add a new buttons param that overrides the existing Standardbuttons. Remember, there's already an ::Ignore button and you can word your dialog in a way that makes sense without using "Install".
- mb->exec() should already return the chosen result if you don't add custom buttons to this generic dialog. So you don't have to check some clickedButton

elad335 replied:
Fixed.

Megamouse commented on rpcs3/rpcs3qt/gui_settings.cpp:
```diff
@@ -225,10 +243,8 @@ bool gui_settings::GetBootConfirmation(QWidget* parent, const gui_save& gui_save
-				"\n<br>For more information read the <a href=\"https://rpcs3.net/quickstart\">quickstart guide</a>."
-				"\nCommercial games do not work without firmware! Do you wish to continue!?");
-			icon = QMessageBox::Critical;
+			message = tr("Installing firmware (PS3UPDAT.PUP) file is required for commercial games."
```
This sentence sounds weird. Idk, maybe improve it somehow.

elad335 replied:
Reworded.

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -847,12 +851,21 @@ void main_window::HandlePupInstallation(QString file_path)
 
+	if (std::string installed = utils::get_firmware_version(); !installed.empty())
+	{
+		if (QMessageBox::question(this, tr("RPCS3 Firmware Installer"), tr("Firmware of version %1 has already been installed.\nOverwrite current installation with %2 version?").arg(qstr(installed), qstr(version_string)),
```
```suggestion
		if (QMessageBox::question(this, tr("RPCS3 Firmware Installer"), tr("Firmware version %1 has already been installed.\nOverwrite current installation with version %2?").arg(qstr(installed), qstr(version_string)),
```

elad335 replied:
Fixed.

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -777,13 +778,16 @@ void main_window::HandlePupInstallation(QString file_path)
 	}
 
-	if (!m_gui_settings->GetBootConfirmation(this))
+	if (!Emu.IsStopped() && !Emu.IsReady())
```
This is technically a regression, as people might also try to install the firmware while in ready state.

elad335 replied:
Fixed, added create_cache optional parameter.

Megamouse commented on rpcs3/rpcs3qt/gui_application.cpp:
```diff
@@ -354,6 +354,11 @@ void gui_application::InitializeCallbacks()
 
+	callbacks.on_install_fw = [this]()
+	{
+		if (m_main_window) m_main_window->InstallPup();
```
There are some real problems here.
- Cancelling the file explorer will simply boot the game instead of handling things gracefully. This looks like a bug from user perspective.
- InstallPup is just semi-blocking. This means you get to choose a PUP file, then the game continues, then a dialog might pop up that asks if you wanna install v.4.85 or whatever, and finally the just-booted game stops and the firmware is installed.
- After installing the firmware, the game doesn't continue.

I'm starting to wonder if you shouldn't just add a debug setting for booting without firmware instead of allowing users to boot without. it and adding a clunky dialog that complicates things for no reason.

elad335 replied:
Fixed.

elad335 commented:
Maybe it was really a bad decision to make firmware-less mode available from GUI, I think it can be made available only when literally all SPRX files are HLEd in firmware settings without the GUI pop-up confirmation.

MSuih commented:
I feel so as well, a normal user will have no need to boot games without firmware and thus this will just create confusion. We can revisit the change if and when HLE is improved so much that it's viable option for commercial games.

elad335 commented:
Fixes #9669

elad335 commented:
Also rewritten the whole thing so its much better on GUI code side.

elad335 commented:
This also means that if the user removed firmware PPU cache, then booted a game, it will compile it whole at startup.

elad335 commented:
* Fix PPU LLVM cache creation lookup in subdirectories. (missed a backslash lmao)

elad335 commented:
* Fixed minor race condition GUI firmware installation cancellation: `progress` variable was incremented non-atomically, possibly dicarding abort status (-1, turns into 0 when incrementing).
* Fixed error handing in firmware installer thread, whenever there's an error abort all execution instead of proceeding normally.

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -245,6 +246,27 @@ QIcon main_window::GetAppIcon()
+	mb->deleteLater();
+	mb->setTextFormat(Qt::RichText);
+
+	mb->addButton(tr("Cancel"), QMessageBox::RejectRole);
```
```suggestion
	mb->addButton(QMessageBox::Cancel);
```

elad335 replied:
Fixed.

Megamouse commented on rpcs3/Emu/System.cpp:
```diff
@@ -1645,6 +1539,32 @@ game_boot_result Emulator::Load(const std::string& title_id, bool add_only, bool
+				// Don't prompt GUI confirmation if there aren't any
+				if (needs_fw && !GetCallbacks().on_missing_fw())
+				{
+					Stop();
```
Probably fixed by neko hotfix

Nekotekina replied:
It's fixed in master.

Nekotekina commented on rpcs3/Emu/Cell/PPUThread.cpp:
```diff
@@ -2019,8 +2034,171 @@ extern void ppu_initialize()
+
+		std::vector<std::string> dir_queue;
+		dir_queue.emplace_back(Emu.GetBoot() + '/');
+		dir_queue.emplace_back(g_cfg.vfs.get_dev_flash() + "sys/external/");
```
Why you scan dev_flash unconditionally?

Nekotekina replied:
Also it seems identical to `firmware_sprx_path` variable.

elad335 replied:
I mentioned it in https://github.com/RPCS3/rpcs3/pull/9664#issuecomment-768990072

elad335 replied:
Also it's for main_window::OnMissingFw() to not call CreateFirmwareCache() which requires terminating current process. (here it's in ready state)
So cache creation has been moved to here instead.

Nekotekina replied:
Anyway, rescanning everything can become very slow even in the case everything is compiled. Need to trade some reliability to the ability to boot at least as fast as before. In other words, find some way to determine the necessity very fast.

Megamouse commented:
Please move as much of the fixes as possible to a new PR that will be reviewed first.
Makes it easier for us to review.

elad335 commented:
* Added support Metal Gear Solid'4 overlays support for precompilation.

Nekotekina commented on rpcs3/Emu/Cell/PPUThread.cpp:
```diff
@@ -2640,3 +2932,5 @@ static void ppu_initialize2(jit_compiler& jit, const ppu_module& module_part, co
 	jit.add(std::move(_module), cache_path);
 #endif // LLVM_AVAILABLE
 }
+
```
Undef at the end of cpp file has no effect

elad335 replied:
Remanant of old code, removed.

Nekotekina commented on rpcs3/Emu/Cell/PPUThread.cpp:
```diff
@@ -2013,17 +2037,278 @@ extern void ppu_initialize()
+					err = obj;
+					break;
+				}
+				case exec_type::ovl:
```
Why letting workers compile OVL if it's just like normal executable can only be compiled sequentially?

elad335 replied:
I think MGS4 can load multiple at once, it depends on the addresses they are using.

elad335 replied:
Also I may shuffle it differently so some threads compile OVL while other PRX.

Nekotekina commented on rpcs3/Emu/Cell/PPUThread.cpp:
```diff
@@ -2013,17 +2037,278 @@ extern void ppu_initialize()
+						return exec_type::prx;
+					}
+
+					if (filename.ends_with(".SELF"sv) || filename.ends_with(".ELF"sv))
```
Or it could be another "main executable"

elad335 replied:
I implemented basic error checking for it in ppu_load_overlay.

elad335 commented:
Rebased pr.

Nekotekina commented on rpcs3/Emu/Cell/PPUThread.cpp:
```diff
@@ -2148,20 +2245,20 @@ extern void ppu_precompile(std::vector<std::string>& dir_queue)
+	shared_mutex sprx_mtx, ovl_mtx;
 
-	named_thread_group workers("SPRX Worker ", utils::get_thread_count(), [&]
+	named_thread_group workers("PRX/OVL Worker ", std::min<u32>(utils::get_thread_count(), ::size32(file_queue)), [&]
```
Keep SPRX Worker please (too long)

elad335 replied:
Fixed.

Nekotekina commented on rpcs3/Emu/Cell/PPUThread.cpp:
```diff
@@ -2099,17 +2133,72 @@ extern void ppu_precompile(std::vector<std::string>& dir_queue)
+
 			// Check .sprx filename
-			if (fmt::to_upper(entry.name).ends_with(".SPRX"))
+			if (upper.ends_with(".SPRX") || upper.ends_with(".PRX"))
```
Why compile .prx or .elf?

elad335 replied:
Removed.

elad335 commented:
For testers: if you have incomplete cache of a game, it won't precompile anything. So you gotta start with a peer game clean cache.
Of course after this pr it won't really be an issue because there shouldn't be such incomplete states . (unless a crash happened while compiling crash such as BSOD)

Nekotekina commented on rpcs3/Emu/Cell/PPUThread.cpp:
```diff
@@ -2064,19 +2071,37 @@ extern void ppu_finalize(const ppu_module& info)
 {
-	std::vector<std::pair<std::string, u64>> file_queue;
-	file_queue.reserve(2000);
+	const std::string firmware_sprx_path = g_cfg.vfs.get_dev_flash() + "sys/external/";
```
Why not vfs::get?

elad335 replied:
Fixed.

Nekotekina commented on rpcs3/Emu/Cell/PPUThread.cpp:
```diff
@@ -2099,17 +2126,72 @@ extern void ppu_precompile(std::vector<std::string>& dir_queue)
-			if (fmt::to_upper(entry.name).ends_with(".SPRX"))
+			if (upper.ends_with(".SPRX"))
 			{
+				// Skip already loaded modules or HLEd ones
```
I only partially understand this part. If they are already loaded, ok. But if they are HLEd, we used to still compile them.

elad335 replied:
I tried to compile them as well, but it messed up their HLE into non-existing LLE (which crashed of course), I think it's a bug/unimplemented feature of ppu_unload_prx.

ZeeWanderer commented:
Cleaned cache folder

as of b11200513
MGS 4 [BLES00246] silently crashes emu during PPU compile
[RPCS3.gz](https://github.com/RPCS3/rpcs3/files/5895697/RPCS3.gz)

Unnamed 5 and Killzone 3 seem to work fine.

elad335 commented:
@ZeeWanderer  A fix was recently merged onto master, retest?

ZeeWanderer commented:
as of 188a358
with the fix, MGS 4 now throws a number of fatal thread termination errors and gets stuck on PPU compilation
[RPCS3.gz](https://github.com/RPCS3/rpcs3/files/5895798/RPCS3.gz)

elad335 commented:
@ZeeWanderer Try now.

elad335 commented:
Moved PPU LLVM stuff onto #9680

ZeeWanderer commented:
retesting is going to take a while
MGS 4 has too much PPU modules.

![image](https://user-images.githubusercontent.com/8377469/106323331-f07ca000-627f-11eb-85aa-d2b5fc94837a.png)

I will edit this comment after it is done.

elad335 commented:
Fixed conflicts.

Megamouse reviewed (changes requested):
A better solution was already discussed

Megamouse commented on rpcs3/rpcs3qt/main_window.h:
```diff
@@ -84,6 +84,9 @@ class main_window : public QMainWindow
 	QIcon GetAppIcon();
+	bool OnMissingFw();
+
+	static bool HandlePupInstallation(main_window* mw, QString file_path = "");
```
all of the main window checks are useless and this shouldn't be a static function.

elad335 replied:
Removed.

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -884,56 +916,83 @@ void main_window::HandlePupInstallation(QString file_path)
+		return false;
+	}
+
+	QString success_str = tr("Successfully installed PS3 firmware (v%1) and LLE Modules!").arg(qstr(version_string));
```
we don't translate log messages, as these are supposed to be readable by developers.
So just copy paste the message without translation to the log message

elad335 replied:
Removed.

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -760,51 +782,53 @@ void main_window::InstallPup(QString file_path)
-	m_gui_settings->SetValue(gui::fd_install_pup, QFileInfo(file_path).path());
 	const std::string path = sstr(file_path);
 
+	const auto critical = [mw](QString str)
```
this can be ommited, as this function cannot and shouldn't be called without a main window.

elad335 replied:
Removed.

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -760,51 +782,53 @@ void main_window::InstallPup(QString file_path)
-	if (!file_path.isEmpty())
-	{
-		// Handle the actual installation with a timeout. Otherwise the source explorer instance is not usable during the following file processing.
-		QTimer::singleShot(0, [this, file_path]()
```
this is not to be changed. we don't want to block the file explorer.

elad335 replied:
Removed.

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -850,29 +874,37 @@ void main_window::HandlePupInstallation(QString file_path)
-	atomic_t<int> progress(0);
+	atomic_t<uint> progress(0);
+
+	do
```
what's the issue with the existing code ?

elad335 replied:
Removed.

elad335 commented:
Updated pr.

Megamouse commented on rpcs3/Emu/System.cpp:
```diff
@@ -1645,9 +1645,36 @@ game_boot_result Emulator::Load(const std::string& title_id, bool add_only, bool
+							// Sanity check because we are using GUI signals for it, could have been installed already
+							if (!fs::is_file(g_cfg.vfs.get_dev_flash() + "sys/external/liblv2.sprx"))
+							{
+								GetCallbacks().on_missing_fw();
```
so this doesn't handle the return value?

Megamouse replied:
You think it's not needed if we abort anyway, right?

elad335 replied:
The current code does not rely on it but in the future this can easily be changed, is what I think.

elad335 commented:
Merge?

Megamouse commented on rpcs3/rpcs3qt/main_window.cpp:
```diff
@@ -249,6 +250,27 @@ QIcon main_window::GetAppIcon()
+{
+	const QString title = tr("Missing Firmware Detected!");
+	const QString message = tr("Commercial games require the firmware (PS3UPDAT.PUP file) to be installed."
+				"\n<br>For information about how to obtain it read the <a href=\"https://rpcs3.net/quickstart\">quickstart guide</a>.");
```
This reads weirdly.
```suggestion
				"\n<br>For information about how to obtain the required firmware read the <a href=\"https://rpcs3.net/quickstart\">quickstart guide</a>.");
```

elad335 replied:
Fixed.

elad335 replied:
Fixed.

Megamouse commented on rpcs3/Emu/System.cpp:
```diff
@@ -1647,9 +1647,36 @@ game_boot_result Emulator::Load(const std::string& title_id, bool add_only, bool
+						CallAfter([this]()
+						{
+							// Sanity check because we are using GUI signals for it, could have been installed already
+							if (!fs::is_file(g_cfg.vfs.get_dev_flash() + "sys/external/liblv2.sprx"))
```
how is anyone supposed to install the firmware in the time it takes to process the last 20 lines?

elad335 replied:
Fixed.

Megamouse commented on rpcs3/Emu/System.cpp:
```diff
@@ -1647,9 +1647,36 @@ game_boot_result Emulator::Load(const std::string& title_id, bool add_only, bool
+
+					for (const auto& lib : g_prx_list)
+					{
+						if (libs.count(std::string(lib.first) + ":lle") || (!lib.second && !libs.count(std::string(lib.first) + ":hle")))
```
isn't the new c++20 thing to use .contains ?
Also, what is lib.second. there's no comment anywhere. not even where it's defined

elad335 replied:
Not sure about compiler support. https://en.cppreference.com/w/cpp/compiler_support
Added comment.

Megamouse replied:
Idk if it matters, but in theory you could use a std::any_of if it supports maps.

elad335 replied:
Done, it supports everything with iterator support.

Megamouse reviewed (approved):
I tested yesterday and it seemed fine enough.
Code shouldn't have degraded.
